src/docfx/build/page/AttributeTransformer.cs

```diff
@@ -4,78 +4,141 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal static class AttributeTransformer
     {
-        public static Func<IEnumerable<DataTypeAttribute>, object, string, object> Transform(
+        private static ThreadLocal<Stack<UidPropertyReference>> t_recursionDetector = new ThreadLocal<Stack<UidPropertyReference>>(() => new Stack<UidPropertyReference>());
+
+        public static Func<IEnumerable<DataTypeAttribute>, object, string, object> TransformSDP(
+            Context context,
+            List<Error> errors,
+            Document file,
+            Action<Document> buildChild)
+        {
+            return Transform;
+
+            object Transform(IEnumerable<DataTypeAttribute> attributes, object value, string jsonPath)
+            {
+                var attribute = attributes.SingleOrDefault(attr => !(attr is XrefPropertyAttribute));
+                return TransformContent(context, errors, attribute, value, file, buildChild);
+            }
+        }
+
+        public static Func<IEnumerable<DataTypeAttribute>, object, string, object> TransformXref(
             Context context,
             List<Error> errors,
             Document file,
             Action<Document> buildChild,
-            JObject extensionData = null)
+            Dictionary<string, Lazy<Func<string, string, Document, Document, JValue>>> extensionData)
         {
             return TransformXrefSpec;
 
             object TransformXrefSpec(IEnumerable<DataTypeAttribute> attributes, object value, string jsonPath)
             {
                 var attribute = attributes.SingleOrDefault(attr => !(attr is XrefPropertyAttribute));
-                var result = TransformContent(attribute, value);
+                extensionData[jsonPath] = new Lazy<Func<string, string, Document, Document, JValue>>(() => Load);
+                return null;
 
-                if (extensionData != null && attributes.Any(attr => attr is XrefPropertyAttribute))
+                JValue Load(string propertyName, string uid, Document referencedFile, Document rootFile)
                 {
-                    extensionData[jsonPath] = new JValue(result);
+                    try
+                    {
+                        if (rootFile != null)
```

Comment:
when is `rootFile` null?

Reply:
No need to check recursion for "Name" property, maybe a flag would be better to identify whether to check recursion or not.

Reply:
I was wrong, we can add check recursion to "Name" as well